### PR TITLE
unix,win: add uv_getenv() and uv_setenv()

### LINKED ISSUE #1198: Req : set/get environment variables
Libuv will [check certain environment variable indirectly](http://docs.libuv.org/en/v1.x/misc.html?highlight=environment#c.uv_os_homedir).  It would be great to have ways of setting/getting arbitrary environment variables.

### PR DESCRIPTION
**Pull request by cjihrig**
These functions are used to create, update, and retrieve environment variables.

Closes #1198

### CODE REVIEW

File: src/win/util.c [hunk 1]
```diff
@@ -1387,3 +1390,146 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+
+  *size = bufsize - 1;
+  return 0;
+}
+
+
+int uv_setenv(const char* name, const char* value) {
+  wchar_t* name_w;
+  wchar_t* value_w;
+  DWORD bufsize;
```
Comment (santigimeno) [inline src/win/util.c]: Should `name_w` and `value_w` be initialized to `NULL` so `out` works correctly on every case?

File: src/win/util.c [hunk 2]
```diff
@@ -1387,3 +1390,146 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+  if (len == 0) {
+    r = GetLastError();
+
+    if (r == ERROR_ENVVAR_NOT_FOUND)
+      return UV_ENOENT;
+
+    return uv_translate_sys_error(r);
+  } else if (len > MAX_ENV_VAR_LENGTH) {
+    /* This should not be possible */
+    return UV_EIO;
```
Comment (santigimeno) [inline src/win/util.c]: If it shouldn't be possible, would an assertion work here?

Comment (cjihrig) [inline src/win/util.c]: An assertion would work. We do this in a few other places though. I don't have a strong preference either way.

File: src/unix/core.c [hunk 3]
```diff
@@ -1240,3 +1240,41 @@ int uv_translate_sys_error(int sys_errno) {
...
+  size_t len;
+
+  if (name == NULL || buffer == NULL || size == NULL || *size == 0)
+    return -EINVAL;
+
+  var = getenv(name);
+
+  if (var == NULL) {
+    return -ENOENT;
+  }
```
Comment (santigimeno) [inline src/unix/core.c]: style: remove braces

File: src/win/util.c (hunk 2 above)
Comment (santigimeno) [inline src/win/util.c]: I'm fine either way too

File: src/unix/core.c [hunk 4]
```diff
@@ -1240,3 +1240,40 @@ int uv_translate_sys_error(int sys_errno) {
...
+
+  len = strlen(var);
+
+  if (len >= *size) {
+    *size = len + 1;
+    return -ENOBUFS;
+  }
+
+  memcpy(buffer, var, len + 1);
+  *size = len;
```
Comment (saghul) [inline src/unix/core.c]: please double check if we do this or include the null byte count in other places.

File: src/win/util.c [hunk 5]
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+    return UV_EINVAL;
+
+  /* Determine the size of the wide character name */
+  bufsize = MultiByteToWideChar(CP_UTF8, 0, name, -1, NULL, 0);
+
+  if (bufsize == 0)
+    return uv_translate_sys_error(GetLastError());
+
+  /* Convert the environment variable name to a wide character string */
+  name_w = (wchar_t*) uv__malloc(sizeof(wchar_t) * bufsize);
```
Comment (saghul) [inline src/win/util.c]: unnecessary cast

File: src/win/util.c (hunk 2 above)
Comment (saghul) [inline src/win/util.c]: I'd go for an assert. If we ever see it blow up IRL we can replace it with another error.

File: src/win/util.c (hunk 1 above)
Comment (saghul) [inline src/win/util.c]: they are, below, right?

File: src/win/util.c [hunk 6]
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+  /* Determine the size of the wide character name */
+  bufsize = MultiByteToWideChar(CP_UTF8, 0, name, -1, NULL, 0);
+
+  if (bufsize == 0) {
+    r = uv_translate_sys_error(GetLastError());
+    goto out;
+  }
+
+  /* Convert the environment variable name to a wide character string */
+  name_w = (wchar_t*) uv__malloc(sizeof(wchar_t) * bufsize);
```
Comment (saghul) [inline src/win/util.c]: unnecessary cast

File: src/win/util.c [hunk 7]
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+  /* Determine the size of the wide character value */
+  bufsize = MultiByteToWideChar(CP_UTF8, 0, value, -1, NULL, 0);
+
+  if (bufsize == 0) {
+    r = uv_translate_sys_error(GetLastError());
+    goto out;
+  }
+
+  /* Convert the environment variable value to a wide character string */
+  value_w = (wchar_t*) uv__malloc(sizeof(wchar_t) * bufsize);
```
Comment (saghul) [inline src/win/util.c]: ditto

File: src/win/util.c [hunk 8]
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+  if (bufsize == 0) {
+    r = uv_translate_sys_error(GetLastError());
+    goto out;
+  }
+
+  if (SetEnvironmentVariableW(name_w, value_w) == 0)
+    r = uv_translate_sys_error(GetLastError());
+
+out:
+  if (name_w)
```
Comment (saghul) [inline src/win/util.c]: no need, free(NULL) is ok

File: src/win/util.c [hunk 9]
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+    uv__free(name_w);
+
+  if (value_w)
```
Comment (saghul) [inline src/win/util.c]: ditto

**Comment by saghul** [comment]
> I'll create a follow up PR for uv_unsetenv() if you're OK with that.

If I may :-) I'd like to have it on this one... that way the feature is complete.

**Comment by saghul** [comment]
Naming bikeshed: should ve have these be `uv_os_setenv` and the like not to overpopulate the `uv_` namespace?

File: src/win/util.c (hunk 1 above)
Comment (santigimeno) [inline src/win/util.c]: Now, they are. IIRC they were not in the initial commit.

File: src/unix/core.c (hunk 4 above)
Comment (cjihrig) [inline src/unix/core.c]: Yea, last year we decided on:

- On `UV_ENOBUFS` set `size` to the required space, including the null.
- On success, set `size` to the string length and null terminate the string.

I don't know if there are functions that still don't follow this, but there were at least a few PRs to migrate the codebase toward this.

File: src/unix/core.c [hunk 10]
```diff
@@ -1240,3 +1240,51 @@ int uv_translate_sys_error(int sys_errno) {
...
+
+  return 0;
+}
+
+
+int uv_os_unsetenv(const char* name) {
+  if (name == NULL)
+    return -EINVAL;
+
+  if (unsetenv(name) != 0)
```
Comment (santigimeno) [inline src/unix/core.c]: I wonder if checking for the `name` being `NULL` is really necessary as `unsetenv()` already does that. The same could be said of `setenv()`.

File: src/win/util.c [hunk 11]
```diff
@@ -1387,3 +1390,180 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+  /* Convert the environment variable name to a wide character string */
+  name_w = uv__malloc(sizeof(wchar_t) * bufsize);
+
+  if (name_w == NULL) {
+    r = UV_ENOMEM;
+    goto out;
+  }
+
+  bufsize = MultiByteToWideChar(CP_UTF8, 0, name, -1, name_w, bufsize);
+
```
Comment (santigimeno) [inline src/win/util.c]: Couldn't this `MultiByteToWideChar` / `malloc` / `MultiByteToWideChar` operation be moved to a separate helper function as it is repeated multiple times?

Comment (cjihrig) [inline src/win/util.c]: I'm in the process of doing that already :-)

**Comment by cjihrig** [comment]
@santigimeno @saghul all existing nits addressed, and added `uv_os_unset()` to this PR. I added a second commit that contains a helper function for the UTF8 to UTF16 conversion on windows.

File: src/win/util.c [hunk 12]
```diff
@@ -1387,3 +1431,113 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+    return r;
+  }
+
+  r = SetEnvironmentVariableW(name_w, value_w);
+  uv__free(name_w);
+  uv__free(value_w);
+
+  if (r == 0)
+    r = uv_translate_sys_error(GetLastError());
+
```
Comment (santigimeno) [inline src/win/util.c]: `return uv_translate_sys_error(GetLastError());` here directly?

File: src/win/util.c [hunk 13]
```diff
@@ -1387,3 +1431,113 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
...
+  int r;
+
+  if (name == NULL)
+    return UV_EINVAL;
+
+  r = uv__convert_utf8_to_utf16(name, -1, &name_w);
+
+  if (r != 0) {
+    return r;
+  }
```
Comment (santigimeno) [inline src/win/util.c]: drop braces

**Comment by cjihrig** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/222/

**Comment by cjihrig** [comment]
Only one test failure (unrelated to this PR). So close to a green CI.

File: docs/src/misc.rst [hunk 14]
```diff
@@ -386,3 +386,38 @@ API
...
+    `value`.
+
+    .. warning::
+        This function is not thread safe.
+
+    .. versionadded:: 1.12.0
+
+.. c:function:: int uv_os_unsetenv(const char* name)
+
+    Deletes the environment variable specified by `name` with. If no such
```
Comment (saghul) [inline docs/src/misc.rst]: s/with//

File: src/unix/core.c (hunk 4 above)
Comment (saghul) [inline src/unix/core.c]: Okidoki!

**Pull request merged**